streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -30,6 +31,11 @@
 class InMemoryTimeOrderedKeyValueBuffer implements TimeOrderedKeyValueBuffer {
     private final Map<Bytes, TimeKey> index = new HashMap<>();
     private final TreeMap<TimeKey, ContextualRecord> sortedMap = new TreeMap<>();
+
+    // track keys that need to be flushed.
+    // keys should be removed before update to maintain update order, rather than insertion order.
+    private final LinkedHashSet<Bytes> dirtyKeys = new LinkedHashSet<>();
```

Comment:
In an effort to limit traffic to the broker, I'm just keeping track of the keys that have been modified since the last flush.

As you'll see in the Store, when we do a flush, *then* we send the most recent state for all these keys to the record collector.

In the long run, I'd like to do some lower-level optimization to support compacting the Producer's buffer for compacted topics, in which case we could drop this optimization here.

But it seemed prudent, given the team's concerns about the performance impact of the buffer's changelog.

Reply:
This makes sense.

Reply:
This isn't a bad idea, but I think it's mitigated by the fact that we clear the set instead of discarding it and creating a new one on flush. This means that it should quickly "settle in" to a size that accomodates the working set and stay there for the duration of the execution. So the cost of growing the set would only be incurred within the first few minutes or hours of execution.

I think...